### DISCUSSION TITLE
enable forget password

### DISCUSSION THREAD

**Discussion by abualmazen** [serenity-is/Serenity#5911] [Help]
hi 
 i try to enable forget password , i use this code 
<system.net>
	<mailSettings>
		<smtp>
			<network defaultCredentials="false" enableSsl="false" host="smtp.gmail.com" port="587" password="*******" userName="newomani.ob@gmail.com" />
		</smtp>
	</mailSettings>
</system.net>
but no email send
it only save email in app_data mail fplder
i am using serenity core

**Comment by VictorTomaili** [comment]
First of all, please use lower case, then configure your email settings. By default for don't send an email by mistake, it's configured for save to directory.

**Reply by LSNicholls** [reply]
Victor, it makes sense that it is configured not to send by mistake.  However, I have the following questions, as I try to enable:

1.  I am not clear about whether "MailingService" plays into this at all -- is this related to the AutoUseQueue value in the [SmtpSettings.cs](https://github.com/serenity-is/common-features/blob/7c60ee7fdaa94e5103e6f89caf1d65dabb64e070/src/Serenity.Extensions/Serenity.Extensions/EmailSender/SmtpSettings.cs) class?  Or is "MailingService" section only used for the email client feature, in which case I would want to disable it?

2.  The sample settings show **PickupPath**-type smtp.  I am assuming based on MailKit that we would use syntax like this as the settings, and it should just work.  I haven't gotten this to work so far, it could be a security setting on my end or something else, so I just want to verify whether MailingService has to be Enabled and AutoUse true for this to work, and to confirm that this SmtpSettings configuration **should** work:

```
// would not hurt to include this in a second,
// commented-out version of the section in the json file, btw:

  "SmtpSettings": {
    "Server": null,
    "Port": 587,
    "SenderName":  "My Name",
    "SenderEmail":  "<my account-user>@<server>.com",
    "Username": "<my-account-user>@<server>.com",
    "Password": "<my-account-password>"
  },
```

**Reply by LSNicholls** [reply]
Further information:

I have written a small test program using MailKit to prove that my settings will indeed send the email successfully.  So, it's not firewall-related that it's not working, and my credentials and mail server, etc, are correct.

Yet, in the Serenity app, the messages continue to go to the App_Data\Mail folder as if for pickup.

BTW I did put MailingService Enabled and AutoUse to true, just in case.  And I put Interval to 1 just in case.  **I don't know if this is necessary at all -- please let me know ??**

The only thing I can think of that is different in my test program is that I used SecureSocketOptions.StartTls, which is required by my server.  I don't know what the right json setting is for this in the appsettings file.  I tried "StartTLS": true,  but it did not work.

**What is the correct way to set up the SmtpSettings json to send email via an external server versus the Pickup Folder syntax?**   I have looked really hard for this on the web, couldn't find anything that showed out to set StartTLS.

And ** is MailingService involved or not? **   This part I am thinking might be Serenity-specific, although you would think the SmtpSettings part would be standard for MailKit.

**Reply by LSNicholls** [reply]
See below, I finally got it, at least for my server.  And to answer my own question, MailingService can be safely disabled.

**Comment by abualmazen** [comment]
hi edited to lower case 
please how i can change default  to send email ?

**Comment by VictorTomaili** [comment]
This is the setting that you need to configure in appsettings.json 
https://github.com/serenity-is/common-features/blob/7c60ee7fdaa94e5103e6f89caf1d65dabb64e070/src/Serenity.Extensions/Serenity.Extensions/EmailSender/SmtpSettings.cs

**Reply by LSNicholls** [reply]
Hi Victor...

>> This is the setting that you need to configure in appsettings.json

OMG.  Looking in https://github.com/serenity-is/common-features/blob/master/src/Serenity.Extensions/Modules/EmailSender/SmtpSettings.cs

This appears to be the magic combination:

```
"SmtpSettings": {
    "Host": "<server>",
    "Port": <port>,
    "SecureSocket": "StartTLS",
    "From": "<from name>",
    "Username": "<user name>",
    "Password": "<pw>"
  },
```

Again, a commented-out second SmtpSettings section (maybe showing the allowable values for SecureSocket???) might be a good idea...

It's working now...

**Comment by elenaaralla** [comment]
I've configured appsettings.json as suggested:

```
"SmtpSettings": {
    "Host": "smtp.gmail.com",
    "Port": "465",
    "SecureSocket": "StartTLS",
    "From": "<from email>",
    "Username": "<username>",
    "Password": "<password>"
  },
```

but emails are only saved in App_Data\Mail; 

I put a try/catch around the send call in AccountController.ForgotPassword.cs;  no error occurs

Debugging I noticed that emailSender.setting are all null or default.

What am I missing?

**Reply by LSNicholls** [reply]
If the emailSender.settings are null or default, when you debug, could there be different settings in appsettings.Development.json?  Just a shot in the dark.

**Reply by elenaaralla** [reply]
Hi,
I tryed using the same settings in all settings file, i.e.  

appsettings.json
appsettings.Development.json
appsettings.Production.json

with no success :(

Thank you!

**Comment by elenaaralla** [comment]
Hi @LSNicholls,

I have still problem :)

Did you configured both web.config and appsettings for mailSender?

If yes, how?

I am not able to send email from serenity and I do not understand why!

When debug, all settings form emailSender are null... I configured them in all appsettings files to be sure, but email are saved on file system and not sent :(

Thank you,
Elena.

**Comment by LSNicholls** [comment]
Hi @elenaaralla,

No, I didn't put anything in web.config.  If the settings are null when you debug, this must mean that there is something wrong with how the  settings are put into the configuration file(s), but I don't know what.  

I hadn't tested this for a while, to be honest, because we're not normally using email from Serenity...  But I just tried now, and it is still working in an app which is in Serenity 6.7.6 -- I guess I should ask which version you are using?

The only other thing I can tell you is that we have Mailing Service disabled (this is another group of settings), we are not queuing anything.  But if, in debug, you are looking in the EmailSender constructor, I would still expect you to see the settings that you put in.

Is it possible that, when you are changing the appsettings*.json files, you are not changing the ones that are actually in use at runtime?  That is all I can think of sorry!

**Comment by elenaaralla** [comment]
Hi @LSNicholls 
I'm using serenity 6.7.1.

When I debug, the mailsender settings are all Null.

What do you mean with:

**Mailing Service disabled (this is another group of settings)**?

Thank you,
Elena.

**Comment by LSNicholls** [comment]
@elenaaralla, in AppSettings.json there is a separate json group of settings which in my file looks like this, notice  Enabled is false and AutoUse is false -- again I don't know if this has anything to do with what you are seeing:
```
 "MailingService": {
      "Enabled": false,
      "AutoUse": false,
      "Interval": 30,
      "BatchSize": 10,
      "LockDuration": 300,
      "RetryLimit": 5
    },
```

>When I debug, the mailsender settings are all Null.

Yes, you said that already.  But I don't know at what point you are seeing this.  So I specified that I thought the settings should be passed in in the constructor and that I would want to look at the values there.  

I am not very good at this, so forgive me if I am not explaining exactly right.  But you haven't said at what point you are looking, what line of code you are on, etc, when the values are null.

And finally, I don't have a way to check this, so I am hoping that you have, and are checking this:

> Is it possible that, when you are changing the appsettings*.json files, you are not changing the ones that are actually in use at runtime?

**Reply by elenaaralla** [reply]
Hi @LSNicholls,

my problem was that SmtpSettings configuration weren't initialized at startup.

Hi added this code to Startup.cs file 

`services.Configure<SmtpSettings>(Configuration.GetSection(SmtpSettings.SectionKey));`

and now i can send emails

**Comment by LSNicholls** [comment]
Hi @elenaaralla thank you for letting everyone know this was the problem.  It is very odd, though, this should have been there in the generated application.  Maybe you inherited this application and somebody else had removed it?

**Reply by elenaaralla** [reply]
Hi @LSNicholls,

I generated the application by myself; 

I followed the step in https://serenity.is/docs/getting_started/installing_serene_from_visual_studio_gallery

I didn't install Basic sample, nor Northwind, but I don't think that this was the problem.

Thank you,
Elena.

**Comment by LSNicholls** [comment]
Wow @elenaaralla I can't explain why that line was missing.   Thanks for letting me know.
